compress/zstd/zstd.go

```diff
@@ -90,52 +102,73 @@ func (r *reader) WriteTo(w io.Writer) (n int64, err error) {
 // NewWriter implements the compress.Codec interface.
 func (c *Codec) NewWriter(w io.Writer) io.WriteCloser {
 	p := new(writer)
-	if cached := c.encPool.Get(); cached == nil {
-		p.enc, p.err = zstd.NewWriter(w, zstd.WithEncoderLevel(c.zstdLevel()))
-		if p.enc != nil {
+	if enc, _ := c.encoderPool.Get().(*encoder); enc == nil {
+		z, err := zstd.NewWriter(w, zstd.WithEncoderLevel(c.zstdLevel()))
+		if err != nil {
+			p.err = err
+		} else {
+			p.enc = &encoder{z}
 			// We need a finalizer because the writer spawns goroutines
 			// that will only be stopped if the Close method is called.
-			runtime.SetFinalizer(p, func(w *writer) { w.Close() })
+			runtime.SetFinalizer(p.enc, (*encoder).finalize)
 		}
 	} else {
-		p.enc = cached.(*zstd.Encoder)
+		p.enc = enc
 		p.enc.Reset(w)
 	}
 	p.c = c
 	return p
 }
 
+type encoder struct {
+	*zstd.Encoder
+}
+
+func (e *encoder) finalize() {
+	e.Close()
+}
+
 type writer struct {
 	c   *Codec
-	enc *zstd.Encoder
+	enc *encoder
 	err error
 }
 
 // Close implements the io.Closer interface.
 func (w *writer) Close() error {
-	if w.enc == nil {
-		return nil // already closed
+	if w.enc != nil {
+		// Close needs to be called to write the end of stream marker and flush
+		// the buffers. The zstd package documents that the encoder is re-usable
+		// fater being closed.
+		err := w.enc.Close()
+		if err != nil {
+			w.err = err
+		}
+		w.enc.Reset(devNull{}) // don't retain the underyling writer
+		w.c.encoderPool.Put(w.enc)
+		w.enc = nil
+		return err
 	}
-	err := w.enc.Close()
-	w.c.encPool.Put(w.enc)
-	w.c = nil
-	w.enc = nil
-	w.err = io.ErrClosedPipe
```

Comment:
how come we store `err` instead of `io.ErrClosedPipe` now?

Reply:
I missed this, I'll fix.